ports/espressif/bindings/espnow/ESPNow.c

```diff
@@ -0,0 +1,876 @@
+
+MP_PROPERTY_GETSET(espnow_buffer_size_obj,
+    (mp_obj_t)&espnow_get_buffer_size_obj,
+    (mp_obj_t)&espnow_set_buffer_size_obj);
+
+//|     phy_rate: int
+//|     """The ESP-NOW physical layer rate."""
+//|
+STATIC mp_obj_t espnow_get_phy_rate(const mp_obj_t self_in) {
+    espnow_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return MP_OBJ_NEW_SMALL_INT(self->phy_rate);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(espnow_get_phy_rate_obj, espnow_get_phy_rate);
+
+STATIC mp_obj_t espnow_set_phy_rate(const mp_obj_t self_in, const mp_obj_t value) {
+    espnow_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    _set_phy_rate(self, mp_obj_get_int(value));
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(espnow_set_phy_rate_obj, espnow_set_phy_rate);
+
+MP_PROPERTY_GETSET(espnow_phy_rate_obj,
+    (mp_obj_t)&espnow_get_phy_rate_obj,
+    (mp_obj_t)&espnow_set_phy_rate_obj);
+
+//|     stats: Tuple[int, int, int, int, int]
+//|     """Provide some useful stats in a `tuple` of
+//|        (tx_packets, tx_responses, tx_failures, rx_packets, rx_failures). (read-only)"""
+//|
+STATIC mp_obj_t espnow_get_stats(mp_obj_t self_in) {
+    espnow_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return MP_OBJ_NEW_TUPLE(
+        mp_obj_new_int(self->tx_packets),
+        mp_obj_new_int(self->tx_responses),
+        mp_obj_new_int(self->tx_failures),
+        mp_obj_new_int(self->rx_packets),
+        mp_obj_new_int(self->rx_failures));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(espnow_get_stats_obj, espnow_get_stats);
+
+MP_PROPERTY_GETTER(espnow_stats_obj,
+    (mp_obj_t)&espnow_get_stats_obj);
+
+// --- Maintaining the peer table and reading RSSI values ---
+
+// We maintain a peers table for several reasons, to:
+// - support monitoring the RSSI values for all peers; and
+// - to return unique bytestrings for each peer which supports more efficient
+//   application memory usage and peer handling.
+
+// Get the RSSI value from the wifi packet header
+static inline int8_t _get_rssi_from_wifi_packet(const uint8_t *msg) {
+    // Warning: Secret magic to get the rssi from the wifi packet header
+    // See espnow.c:espnow_recv_cb() at https://github.com/espressif/esp-now/
+    // In the wifi packet the msg comes after a wifi_promiscuous_pkt_t
+    // and a espnow_frame_format_t.
+    // Backtrack to get a pointer to the wifi_promiscuous_pkt_t.
+    #define SIZEOF_ESPNOW_FRAME_FORMAT 39
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align"
+    wifi_promiscuous_pkt_t *wifi_packet = (wifi_promiscuous_pkt_t *)(
+        msg - SIZEOF_ESPNOW_FRAME_FORMAT - sizeof(wifi_promiscuous_pkt_t));
+    #pragma GCC diagnostic pop
+    return wifi_packet->rx_ctrl.rssi;
+}
+
+// Lookup a peer in the peers table and return a reference to the item in the peers_table.
+// Add peer to the table if it is not found (may alloc memory). Will not return NULL.
+static mp_map_elem_t *_lookup_add_peer(espnow_obj_t *self, const uint8_t *peer) {
```

Comment:
I'd prefer to split the implementation code into separate files in common-hal similar to `espulp`. That will simplify this file a bunch.